client/src/cmdlfem4x70.c

```diff
@@ -240,11 +240,76 @@ int CmdEM4x70Unlock(const char *Cmd) {
     return PM3_ESOFT;
 }
 
+int CmdEM4x70Auth(const char *Cmd) {
+
+    // Authenticate transponder
+    // Send 56-bit random number + pre-computed f(rnd, k) to transponder.
+    // Transponder will respond with a response
+    em4x70_data_t etd = {0};
+
+    CLIParserContext *ctx;
+
+    CLIParserInit(&ctx, "lf em 4x70 auth",
+                  "Authenticate against an EM4x70 by sending random number (RN) and F(RN)\n"
+                  "  If F(RN) is incorrect based on the tag crypt key, the tag will not respond",
+                  "lf em 4x70 auth -r 11223344556677 -f 11223344\n"
+                 );
+
+    void *argtable[] = {
+        arg_param_begin,
+        arg_lit0("x", "par", "Add parity bit when sending commands"),
+        arg_str1("r", "rnd", "<hex>", "Random 56-bit"),
+        arg_str1("f", "frn", "<hex>", "F(RN) 28-bit as 4 hex bytes"),
+        arg_param_end
+    };
+
+    CLIExecWithReturn(ctx, Cmd, argtable, true);
+
+    etd.parity = arg_get_lit(ctx, 1);
+
+    int rnd_len = 7;
+    CLIGetHexWithReturn(ctx, 2, etd.rnd, &rnd_len);
+
+    int frnd_len = 4;
+    CLIGetHexWithReturn(ctx, 3, etd.frnd, &frnd_len);
+
+    CLIParserFree(ctx);
+
+    if (rnd_len != 7) {
+        PrintAndLogEx(FAILED, "Random number length must be 7 bytes instead of %d", rnd_len);
+        return PM3_EINVARG;
+    }
+
+    if (frnd_len != 4) {
+        PrintAndLogEx(FAILED, "F(RN) length must be 4 bytes instead of %d", frnd_len);
+        return PM3_EINVARG;
+    }
+
+    clearCommandBuffer();
+    SendCommandNG(CMD_LF_EM4X70_AUTH, (uint8_t *)&etd, sizeof(etd));
+
+    PacketResponseNG resp;
+    if (!WaitForResponseTimeout(CMD_LF_EM4X70_AUTH, &resp, TIMEOUT)) {
+        PrintAndLogEx(WARNING, "Timeout while waiting for reply.");
```

Comment:
How is it set up on device on failure?  or timeouts,    does the device drop the field?  
because if not,  this will lead the device to become unresponsive if user starts issuing new commands when this one timeouts.

Reply:
I was trying to follow what the EM4x50 did where on-device there are timeouts during all the loops and limits setup to prevent getting stuck. When we fail to find a tag or perform the command it should call lf_finalize() before returning. I will review the on-device code for any potential areas it could get stuck (loops without a timeout).

So far no commands on the device should hit the client timeout, the device should be timing out first and return "Failed".

In a previous PR when adding write support, I fixed an issue where the receive function could get stuck if the tag is removed.

Reply:
good,   or we need to loop in the client if it is enforced that the device must send an answer to this command.

Reply:
All commands must return a success or failure response. So if there is no response it should be considered a bug in the em4x70 device code.

If I switch to "WaitForResponse" without the timeout, there will be a warning printed that pressing the button will cancel the operation, which isn't true in this case. I could perhaps update the error message to indicate the device may now be unresponsive?

I'm not sure the best option, in both cases I think the device would be unresponsive, but that could be true of a bug anywhere on the device I believe. I believe the watchdog would restart the device in this case?

Reply:
making it fail safe on device side,  so it doesn't get stuff on device side in a loop.    
We already have a timeout on client side now,

Reply:
Sounds good, that should be how it works now. The device-side loops do have timeouts that should then cause us to return an error in the case of no tag or weak tag signal etc.